pkg/util/atomicfs/fsync.go

```diff
@@ -3,44 +3,70 @@
 package atomicfs
 
 import (
-	"bytes"
 	"io"
 	"os"
-	"path"
+	"path/filepath"
 
 	"github.com/grafana/dskit/multierror"
 )
 
-// CreateFile creates a file in the filePath, write the data into the file and then execute
-// fsync operation to make sure the file and its content are stored atomically. If the file already
-// exists, it will be overwritten.
-func CreateFile(filePath string, data io.Reader) error {
-	// Write the file, fsync it, then fsync the containing directory in order to guarantee
-	// it is persisted to disk. From https://man7.org/linux/man-pages/man2/fsync.2.html
-	//
-	// > Calling fsync() does not necessarily ensure that the entry in the
-	// > directory containing the file has also reached disk.  For that an
-	// > explicit fsync() on a file descriptor for the directory is also
-	// > needed.
-	file, err := os.Create(filePath)
+// Create creates a new file at a temporary path that will be renamed to the
+// supplied path on close, ensuring all data and the containing directory have
+// been fsynced to disk.
+func Create(path string) (*File, error) {
+	// Important that the temporary file is in the same directory as the final file.
```

Comment:
(nit) I would expand on why it's important

```suggestion
		// The process might not have permissions to write to the system's tempory directory, thus it is
		// important that we create the temporary file in the same directory as the final file.
```

Reply:
and maybe move the comment on the godoc, not that we generate godocs